File: app/src/main/java/com/example/neon/deardiary/diaryquery/DiaryQueryFragment.java
package com.example.neon.deardiary.diaryquery;

import android.app.Fragment;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ListView;

import com.example.neon.deardiary.R;
import com.example.neon.deardiary.data.Diary;
import com.example.neon.deardiary.diaryedit.DiaryEditActivity;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnItemClick;
import butterknife.Unbinder;


public class DiaryQueryFragment extends Fragment implements DiaryQueryContract.View {

    private static final String TAG = "DiaryQueryFragment";

    @BindView(R.id.search_result_list)
    ListView mLvResult;
    @BindView(R.id.searchKey)
    EditText mEtQuery;

    private DiaryQueryContract.Presenter mPresenter;
    private DiaryQueryAdapter mAdapter;
    private Unbinder mUnbinder;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_diary_query, container, false);
        mUnbinder = ButterKnife.bind(this, root);
        mLvResult.setDividerHeight(0);
        mEtQuery.addTextChangedListener(new TextWatcher() {

            @Override
            public void afterTextChanged(Editable s) {
                String keyword = s.toString();
                beginQuery(keyword);
            }

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }
        });
        return root;
    }

    @OnItemClick(R.id.search_result_list)
    public void onItemClick(int position){
        Diary diary = (Diary) mAdapter.getItem(position);
        Bundle bundle = new Bundle();
        bundle.putSerializable(DiaryEditActivity.DATA_DIARY, diary);
        Intent intent = new Intent(getActivity(), DiaryEditActivity.class);
        intent.putExtras(bundle);
        startActivity(intent);

    }

    @OnClick(R.id.search_close_btn)
    public void close(){
        getActivity().finish();
    }

    private void beginQuery(String keyword) {
        if (!"".equals(keyword)) {
            mLvResult.setVisibility(View.VISIBLE);
            if (mAdapter == null) {
                mAdapter = new DiaryQueryAdapter(getActivity());
                mLvResult.setAdapter(mAdapter);
            }
            mPresenter.diaryQueryByKeyword(keyword);
        } else {
            mLvResult.setVisibility(View.INVISIBLE);
        }
    }

    @Override
    public void setPresenter(DiaryQueryContract.Presenter presenter) {
        mPresenter = presenter;
    }

    @Override
    public void onResume() {
        super.onResume();
        beginQuery(mEtQuery.getText().toString());
    }

    @Override
    public void querySuccessfully(ArrayList<Diary> diaries) {
        mAdapter.setData(diaries);
    }

    @Override
    public void queryFailed() {
        Log.e(TAG, "queryFailed: 查询失败");
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mUnbinder.unbind();
    }
}

File: app/src/main/java/com/example/neon/deardiary/diaryquery/DiaryQueryFragment.kt
package com.example.neon.deardiary.diaryquery

import android.app.Fragment
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import butterknife.Unbinder
import com.example.neon.deardiary.R
import com.example.neon.deardiary.data.Diary
import com.example.neon.deardiary.utils.LogUtil
import java.util.ArrayList

internal class DiaryQueryFragment : Fragment(), DiaryQueryContract.View {
    @BindView(R.id.search_result_list)
    lateinit var mLvResult: RecyclerView
    @BindView(R.id.searchKey)
    lateinit var mEtQuery: EditText

    private lateinit var mPresenter: DiaryQueryContract.Presenter
    private lateinit var mAdapter: DiaryQueryAdapter
    private lateinit var mUnBinder: Unbinder

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val root = inflater.inflate(R.layout.fragment_diary_query, container, false)
        mUnBinder = ButterKnife.bind(this, root)
        mAdapter = DiaryQueryAdapter(activity)
        mLvResult.adapter = mAdapter
        mLvResult.layoutManager = LinearLayoutManager(activity)
        mEtQuery.addTextChangedListener(object : TextWatcher {

            override fun afterTextChanged(s: Editable) {
                val keyword = s.toString()
                beginQuery(keyword)
            }

            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
        })
        return root
    }

    @OnClick(R.id.search_close_btn)
    fun close() {
        activity.finish()
    }

    private fun beginQuery(keyword: String) {
        if ("" != keyword) {
            mLvResult.visibility = View.VISIBLE
            mPresenter.diaryQueryByKeyword(keyword)
        } else {
            mLvResult.visibility = View.INVISIBLE
        }
    }

    override fun setPresenter(presenter: DiaryQueryContract.Presenter) {
        mPresenter = presenter
    }

    override fun onResume() {
        super.onResume()
        beginQuery(mEtQuery.text.toString())
    }

    override fun querySuccessfully(diaries: ArrayList<Diary>) {
        mAdapter.setData(diaries)
    }

    override fun queryFailed() {
        LogUtil.e(TAG, "queryFailed: 查询失败")
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mUnBinder.unbind()
    }

    companion object {
        private val TAG = "DiaryQueryFragment"
    }
}
